### PR TITLE
Shutdown unsupported bloop

### PR DESCRIPTION
**Pull request by tpasternak**

### CODE REVIEW

File: build.sc [hunk 1]
```diff
@@ -350,7 +350,8 @@ trait CliIntegrationBase extends SbtModule with ScalaCliPublishModule with HasTe
            |  def scala213 = "${Scala.scala213}"
            |  def scala3   = "${Scala.scala3}"
            |  def defaultScala = "${Scala.defaultUser}"
-           |
+           |  def bloopVersion = "${Deps.bloopConfig.dep.version}"
+           |  def oldBloopVersion = "${Deps.oldBloopConfig.dep.version}"
```
Comment (alexarchambault) [inline build.sc]: Do we need to write the old bloop version here? I think it's fine to hardcode it in the tests…

File: modules/bloop-rifle/src/main/scala/scala/build/blooprifle/BloopRifle.scala [hunk 2]
```diff
@@ -172,4 +184,51 @@ object BloopRifle {
...
+      workdir,
+      nullInputStream(),
+      bufferedOStream,
+      nullOutputStream(),
+      logger
+    )
+    if (result == 0) {
+      val currentBloopVersion =
+        extractVersionFromBloopAbout(new String(bufferedOStream.toByteArray())).map(Version(_))
+      val minimumBloopVersion = Version(Constants.bloopVersion)
```
Comment (alexarchambault) [inline modules/bloop-rifle/src/main/scala/scala/build/blooprifle/BloopRifle.scala]: The minimum version should be put in `BloopRifleConfig`, with type `Option[String]` say. And no checks would be run if its value is `None`.

Comment (alexarchambault) [reply modules/bloop-rifle/src/main/scala/scala/build/blooprifle/BloopRifle.scala]: Actually, we should just put a field like
```scala
acceptVersion: Option[String => Boolean]
```
accepting or discarding the Bloop version. That way, we wouldn't have to add new dependencies to the `bloop-rifle` module.

That field can default to `None`, and we can change that value around [here](https://github.com/VirtusLab/scala-cli/blob/130e0d763f9b146239304db6f35c6520fabc8a12/modules/cli/src/main/scala/scala/cli/commands/SharedCompilationServerOptions.scala#L164).

Comment (alexarchambault) [reply modules/bloop-rifle/src/main/scala/scala/build/blooprifle/BloopRifle.scala]: For version comparison, coursier should be around there. I think something like this should work:
```scala
import coursier.core.Version
Version("1.2.3").compareTo(Version("1.4.5")) < 0
```

File: modules/bloop-rifle/src/main/scala/scala/build/blooprifle/Version.scala [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package scala.build.blooprifle
+import org.apache.maven.artifact.versioning.ComparableVersion
+case class Version(v: String) {
+  def <(rhs: Version): Boolean = {
+    new ComparableVersion(v).compareTo(new ComparableVersion(rhs.v)) < 0
+  }
```
Comment (alexarchambault) [inline modules/bloop-rifle/src/main/scala/scala/build/blooprifle/Version.scala]: We shouldn't need this, see my comment above (comparison can be done from the `cli` module, using `coursier.core.Version`).

File: modules/bloop-rifle/src/main/scala/scala/build/blooprifle/internal/Operations.scala [hunk 4]
```diff
@@ -327,4 +333,43 @@ object Operations {
...
+        logger.nailgunLogger,
+        stop0,
+        interactiveSession = false
+      )
+    }
+
+  }
+
+  def timeout[T](duration: Duration)(body: => T) = {
+    import scala.concurrent.ExecutionContext.Implicits.global
```
Comment (alexarchambault) [inline modules/bloop-rifle/src/main/scala/scala/build/blooprifle/internal/Operations.scala]: We shouldn't use the default `ExecutionContext` for that (it's rarely a good idea to use this one, it's assumptions sometimes change from a Scala version to another, …)

Some `Operations` methods already accept a `ScheduledExecutorService`. Code along those lines ought to work:
```scala
val p = Promise[Int]()
scheduler.schedule { () =>
  val retCode = nailgunClient.run(…)
  p.tryComplete(Success(retCode))
}
scheduler.schedule(() => { p.tryComplete(Failure(new TimeoutException(…)))  }, 5, MINUTES)
```
(We could even keep the `scheduler.*` calls values, and try to cancel the other when one completes.)

File: project/deps.sc [hunk 5]
```diff
@@ -42,6 +44,7 @@ object Deps {
   def jimfs             = ivy"com.google.jimfs:jimfs:1.2"
   def jniUtils          = ivy"io.get-coursier.jniutils:windows-jni-utils:0.3.1"
   def macroParadise     = ivy"org.scalamacros:::paradise:2.1.1"
+  def mavenArtifact     = ivy"org.apache.maven:maven-artifact:3.8.2"
```
Comment (alexarchambault) [inline project/deps.sc]: We shouldn't need that one, see above.

File: project/deps.sc [hunk 6]
```diff
@@ -27,8 +27,10 @@ object Deps {
     def scalaNative   = "0.4.0"
     def scalaPackager = "0.1.21"
   }
-  def ammonite          = ivy"com.lihaoyi:::ammonite:2.4.0-20-f3d8171f"
-  def asm               = ivy"org.ow2.asm:asm:9.1"
+  def ammonite = ivy"com.lihaoyi:::ammonite:2.4.0-20-f3d8171f"
+  def asm      = ivy"org.ow2.asm:asm:9.1"
+  def oldBloopConfig =
```
Comment (alexarchambault) [inline project/deps.sc]: We shouldn't need that one, we can hardcode that version in the tests.

File: modules/bloop-rifle/src/main/scala/scala/build/blooprifle/internal/Operations.scala (hunk 4 above)
Comment (tpasternak) [reply modules/bloop-rifle/src/main/scala/scala/build/blooprifle/internal/Operations.scala]: Done

File: modules/bloop-rifle/src/main/scala/scala/build/blooprifle/Version.scala (hunk 3 above)
Comment (tpasternak) [reply modules/bloop-rifle/src/main/scala/scala/build/blooprifle/Version.scala]: I tried, but adding coursier to BloopRifle dependencies causes some resolution errors, unfortunately

File: modules/bloop-rifle/src/main/scala/scala/build/blooprifle/BloopRifle.scala (hunk 2 above)
Comment (tpasternak) [reply modules/bloop-rifle/src/main/scala/scala/build/blooprifle/BloopRifle.scala]: Commented here https://github.com/VirtusLab/scala-cli/pull/145#discussion_r717327973

File: build.sc (hunk 1 above)
Comment (tpasternak) [reply build.sc]: Done

File: modules/bloop-rifle/src/main/scala/scala/build/blooprifle/Version.scala (hunk 3 above)
Comment (tpasternak) [reply modules/bloop-rifle/src/main/scala/scala/build/blooprifle/Version.scala]: For some reason guava is resolved to a different version, so we get:
```
Exception in thread "main" java.lang.NoSuchMethodError: com.google.common.base.Preconditions.checkArgument(ZLjava/lang/String;J)V
	at com.google.common.jimfs.WatchServiceConfiguration$PollingConfig.<init>(WatchServiceConfiguration.java:61)
	at com.google.common.jimfs.WatchServiceConfiguration$PollingConfig.<init>(WatchServiceConfiguration.java:55)
	at com.google.common.jimfs.WatchServiceConfiguration.polling(WatchServiceConfiguration.java:44)
	at com.google.common.jimfs.WatchServiceConfiguration.<clinit>(WatchServiceConfiguration.java:35)
	at com.google.common.jimfs.Configuration$Builder.<init>(Configuration.java:346)
	at com.google.common.jimfs.Configuration$Builder.<init>(Configuration.java:318)
	at com.google.common.jimfs.Configuration.builder(Configuration.java:218)
	at com.google.common.jimfs.Configuration$UnixHolder.<clinit>(Configuration.java:90)
	at com.google.common.jimfs.Configuration.unix(Configuration.java:85)
	at com.google.common.jimfs.Configuration.forCurrentPlatform(Configuration.java:212)
	at com.google.common.jimfs.Jimfs.newFileSystem(Jimfs.java:114)
	at com.google.common.jimfs.Jimfs.newFileSystem(Jimfs.java:97)
	at org.scalajs.jsenv.nodejs.ComRun$.setupFile(ComSupport.scala:247)
	at org.scalajs.jsenv.nodejs.ComRun$.start(ComSupport.scala:214)
	at org.scalajs.jsenv.nodejs.NodeJSEnv.startWithCom(NodeJSEnv.scala:46)
	at org.scalajs.testing.adapter.JSEnvRPC.<init>(JSEnvRPC.scala:25)
	at org.scalajs.testing.adapter.TestAdapter.startManagedRunner(TestAdapter.scala:129)
	at org.scalajs.testing.adapter.TestAdapter.$anonfun$getRunnerForThread$1(TestAdapter.scala:120)
	at scala.collection.concurrent.TrieMap.getOrElseUpdate(TrieMap.scala:895)
	at org.scalajs.testing.adapter.TestAdapter.getRunnerForThread(TestAdapter.scala:120)
	at org.scalajs.testing.adapter.TestAdapter.loadFrameworks(TestAdapter.scala:56)
	at scala.build.internal.Runner$.testJs(Runner.scala:247)
	at scala.cli.commands.Test$.$anonfun$testOnce$1(Test.scala:81)
	at scala.cli.commands.Test$.$anonfun$testOnce$1$adapted(Test.scala:76)
	at scala.cli.commands.Run$.withLinkedJs(Run.scala:184)
	at scala.cli.commands.Test$.testOnce(Test.scala:76)
	at scala.cli.commands.Test$.maybeTest$1(Test.scala:30)
	at scala.cli.commands.Test$.run(Test.scala:57)
	at scala.cli.commands.Test$.run(Test.scala:9)
	at caseapp.core.app.CaseApp.main(CaseApp.scala:150)
	at caseapp.core.app.CommandsEntryPoint.main(CommandsEntryPoint.scala:112)
	at scala.cli.ScalaCli$.main(ScalaCli.scala:77)
	at scala.cli.ScalaCli.main(ScalaCli.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at coursier.bootstrap.launcher.a.a(Unknown Source)
	at coursier.bootstrap.launcher.Launcher.main(Unknown Source)
```

**Review by romanowski — APPROVED** [review]

**Comment by alexarchambault** [comment]
I pushed some extra commit, fixing minor stuff, so that this can be merged earlier.

I just understood that during the tests, you actually use both an older and a newer version of Bloop, right? (I had missed the newer one.) I'm not sure we should assume a newer one exists. Ideally, we'd like to always depend on the latest Bloop version.

**Comment by tpasternak** [comment]
Yes, I only wanted to to clearly show in the tests that the version newer than the default one is not killed by scala-cli. A newer version may be spawned by an other tool, like metals, even if we don't bump it

**Comment by tpasternak** [comment]
Can I merge it now?

**Review by alexarchambault — APPROVED** [review]
> Can I merge it now?

Squash-and-merge, yes.

**Pull request merged**